src/pages/home/report/ReportActionCompose.js

```diff
@@ -282,11 +284,21 @@ class ReportActionCompose extends React.Component {
 
     onSelectionChange(e) {
         LayoutAnimation.configureNext(LayoutAnimation.create(50, LayoutAnimation.Types.easeInEaseOut, LayoutAnimation.Properties.opacity));
-        this.setState({selection: e.nativeEvent.selection});
-        if (!this.state.value || e.nativeEvent.selection.end < 1) {
-            this.resetSuggestions();
+
+        if (e) {
+            this.setState({selection: e.nativeEvent.selection});
+            if (!this.state.value || e.nativeEvent.selection.end < 1) {
+                this.resetSuggestions();
+                this.shouldBlockSuggestionsCalc = false;
+                return;
+            }
+        }
+
+        if (this.shouldBlockSuggestionsCalc) {
+            this.shouldBlockSuggestionsCalc = false;
             return;
         }
+
```

Comment:
If you want me to remove it I can, but keep in mind that after I do the issue I mentioned will occur for mWeb, desktop, macOS browsers. Is that fine?

Reply:
Yes

Reply:
This is from `main`.

https://github.com/Expensify/App/assets/60378235/b37d18e3-87b3-4847-bdb8-6bfb11dd0cbe

For this PR (without the change in question).

https://github.com/Expensify/App/assets/60378235/d5af01a7-0eb6-4793-b43a-88eb9d72f069

Reply:
Sure, if you're testing my code don't forget to comment this line.

https://github.com/therealsujitk/expensify-app/blob/39d7ca24fd8c28f0a9fa5ef998703873432f0d7f/src/pages/home/report/ReportActionCompose.js#L1088

Reply:
> Theoretically, this PR just moved code for emoji part (no logic changes) so shouldn't be affected

Yeah that's why I kept testing to see what was wrong, and then I realised it was race conditions. The previous check was inside the `calculateEmojiSuggestion` function which happened to be called after the `onClick` which was setting the variable to `false`.

Reply:
> Can you also test again main? (not staging/production)

Tested, works the same as production.

Reply:
#### This is what happens on Chrome.
1. User clicks on add attachment, `shouldBlockSuggestionsCalc` is set to `true`.
2. The modal is closed and the `onSelectionChange` function is called which sets `shouldBlockSuggestionsCalc` to `false`.
3. The picker then opens, once it's closed and the user clicks on the composer, `onSelectionChange` is called again and it shows the suggestion list since `shouldBlockSuggestionsCalc` is `false`. After this `onClick` is called which sets `shouldBlockSuggestionsCalc` to `false`.

#### This is what happens of mWeb.
1. User clicks on add attachment, `shouldBlockSuggestionsCalc` is set to `true`.
2. The modal is closed, but the `onSelectionChange` function isn't called.
3. 
    - The picker then opens, once it's closed and the user clicks on the composer, `onSelectionChange` is called, however `shouldBlockSuggestionsCalc` is set to `false` only now. 
    - After this `onClick` is called which also sets `shouldBlockSuggestionsCalc` is set to `false`.
5. The suggestions list is only displayed after `onSelectionChange` is called again.

##### Why was it working previously?

The check for `shouldBlockEmojiCalc` happened inside the `calculateEmojiSuggestion`, so what happened was this function happened to be called after the `onClick` took place (In step 3) and so `shouldBlockSuggestionsCalc` was already `false`. (A race condition due to async)

I hope it's clear now.

Reply:
This same is probably happening to you right now which is why it's not reproducible to you.

Reply:
No, that has nothing to do with it. Event functions are async, so there's a possibility of them being executed at the same time, in this case `onSelectionChange` and `onClick`.

Reply:
One way to fix this is to use `onMouseDown` since that is triggered before `onSelectionChange` there's a higher chance it'll execute first (`onClick` is triggered after `onSelectionChange`) but this won't solve the problem on macOS browsers.

Reply:
Ok so previously this is what happened, the check for `shouldBlockEmojiCalc` happened a little later in `calculateEmojiSuggestion` by that time it becomes `false`.

```
Thread - 1                                          Thread - 2
.                                                   .
.                                                   .
.                                                   .
.                                                   shouldBlockEmojiCalc = false;
.
.
.
shouldBlockEmojiCalc // False
```

Now, this is what is happening, we are checking for `shouldBlockSuggestionsCalc` earlier than before and the value is not set to `false` yet.

```
Thread - 1                                          Thread - 2
.                                                   .
shouldBlockSuggectionsCalc // True                  .
                                                    .
                                                    shouldBlockSuggectionsCalc = false;
```

`Thread - 1` is when `onSelectionChange` is called, `Thread - 2` is when `onClick` is called. This behaviour is unpredictable and it depends on the system.

Reply:
OHHH!! That makes so much sense :rofl:.

Reply:
Ok this will solve the issue for mWeb. What about Safari desktop, Do we leave it as is? It's reproducible in production.

Reply:
Note that there's still a very *very* small chance this issue will show up in mWeb, we're basically fixing this with a timer.

Reply:
Understood I'll make the changes! I'd prefer if we create a separate issue for this and don't just forget about it because it seems like bad code. I *probably* won't work on it thought, this has given me quite a headache :sweat_smile:.

Reply:
> ok, let's put that code at the top of both functions

Ah, we can't do this :disappointed:. The `calculateEmojiSuggestion` function is first called when `shouldBlockSuggestionsCalc` is `true`. It sets it to `false` and returns, immediately after `calculateMentionSuggestion` is called which is set to `false` because of `calculateEmojiSuggestion` so it doesn't return even though it should.

To solve this we'll need two variables `shouldBlockEmmojiCalc` & `shouldBlockMentionCalc`. Should I proceed?